### PR TITLE
Add exponential delay between authentications for the same Student

### PR DESCRIPTION
**Pull request by sladomic**

### CODE REVIEW

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java [hunk 1]
```diff
@@ -55,6 +56,7 @@
     private GifImageView authenticationAnimation;
     private boolean recognitionThreadStarted;
     private boolean activityStopped;
+    AuthenticationEventDao authenticationEventDao;
```
Comment (jo-elimu) [inline app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java]: Add `private`

File: app/src/main/java/org/literacyapp/authentication/AuthenticationHelper.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.Calendar;
+import java.util.List;
+
+/**
+ * Created by sladomic on 30.12.16.
+ */
+
+public class AuthenticationHelper {
+    private static final double[] countSteps = { 0, 5, 10, 15, 20 };
+    private static final double[] exponentialDelayStepsInMinutes = { 30, 60, 120, 240, 360 };
```
Comment (jo-elimu) [inline app/src/main/java/org/literacyapp/authentication/AuthenticationHelper.java]: Will 6 hours be the maximum? What if only one Student is using the same tablet during the 18 months? Would it not be too annoying to do the authentication every 6 hours? Maybe we should increase the maximum to something like 5760 or 11520 minutes?

**Review by jo-elimu — APPROVED** [review]

File: app/src/main/java/org/literacyapp/authentication/AuthenticationHelper.java (hunk 2 above)
Comment (sladomic) [reply app/src/main/java/org/literacyapp/authentication/AuthenticationHelper.java]: Hmm not an easy question. I thought at least once per day or so but I'm not sure. You suggest the maximum should be 4 or 8 days?

**Comment by sladomic** [comment]
I just saw that I didn't change the behaviour after reboot. So at the moment it would appear after every reboot and then again reschedule with an exponentialDelay. Shall I change it so, that after the reboot also the exponentialDelay is used or should after a reboot the authentication come up?

File: app/src/main/java/org/literacyapp/authentication/AuthenticationHelper.java (hunk 2 above)
Comment (jo-elimu) [reply app/src/main/java/org/literacyapp/authentication/AuthenticationHelper.java]: Another question is this: how long time should it take to reach the maximum delay? Now it might take 30 + 60 + 120 + 240 + 360 = 810 minutes for one Student to reach the 360 minute delay, right? All in just one day. So maybe it's too aggressive to increase the delay so quickly, especially since it's likely to see different Students using a tablet over a period of 2 days. Tablet switching is more probable to happen each morning than during the day.

How about this: What if we follow the same formula as above, but we only increase the delay if the previous delay adjustment was performed the day before or earlier (not during the same day)? This would require storing a timestamp for the last time the delay was adjusted.

In other words: a longer delay (like 2 or 4 days) might make more sense in the case where only one person is using the same tablet, but it should require several days of 1-person usage before reaching that delay, e.g. at least 2 weeks of usage. Perhaps require at least 1 week of 1-person usage before introducing 1-day delay.

**Comment by jo-elimu** [comment]
> I just saw that I didn't change the behaviour after reboot. So at the moment it would appear after every reboot and then again reschedule with an exponentialDelay. Shall I change it so, that after the reboot also the exponentialDelay is used or should after a reboot the authentication come up?

@sladomic I think it makes sense to have the authentication come up after each reboot, in case the device ran out of battery and had to be charged. I'll add tracking of each device reboot (#311) so that we'll learn more about what the typical pattern of device reboots looks like.

File: app/src/main/java/org/literacyapp/authentication/AuthenticationHelper.java (hunk 2 above)
Comment (sladomic) [reply app/src/main/java/org/literacyapp/authentication/AuthenticationHelper.java]: At the moment it would be 30 + 30 + 30 + 30 + 30 + 60 + 60 + 60 + 60 + 60 + 120 + 120 + 120 + 120 + 120 + 240 + 240 + 240 + 240 + 240 + 360 + 360 + 360 + 360 + 360 = 4050 minutes
